src/main/scala/io/iohk/ethereum/merklepatriciatrie/Node.scala

```diff
@@ -0,0 +1,102 @@
+package io.iohk.ethereum.merklepatriciatrie
+
+import io.iohk.ethereum.rlp.{decode => decodeRLP, encode => encodeRLP, _}
+import io.iohk.ethereum.merklepatriciatrie.MerklePatriciaTrie.HashFn
+
+/**
+  * Trie elements
+  */
+private[merklepatriciatrie] sealed trait Node {
+
+  import MerklePatriciaTrie._
+
+  private var hashedCache: Option[Array[Byte]] = None
```

Comment:
@adamsmo let me see if I can work on that but i'm not sure if it might work as nodes are also created while decoding (and decoder is used as an implicit created within companion object).

Also, let me rephrase. `Node` objects are only created while doing a put or remove operations. They are not part of the Trie state (at least in the form of `Node`) nor shared between operations. This might (and i'm not sure if it will) be a problem if we decide by any reason keep cached in-memory versions of MPT (for example to speedup blocks processing). 
In terms of thread safety, we may need not to allow multiple `put` and `remove` operations from different threads, it's not about nodes. Does this make sense?